jwst/ramp_fitting/ramp_fit.py

```diff
@@ -549,79 +537,332 @@ def ols_ramp_fit(data, err, groupdq, inpixeldq, buffsize, save_opt, readnoise_2d
+    else:
+        opt_slope = None
+        opt_sigslope = None
+        opt_var_poisson = None
+        opt_var_rnoise = None
+        opt_yint = None
+        opt_sigyint = None
+        opt_pedestal = None
+        opt_weights = None
+        opt_crmag = None
+        actual_segments = 0
+        actual_CRs = 0
 
-            first_gdq = groupdq[:, 0, :, :]
+    return new_model.data, new_model.dq, new_model.var_poisson, new_model.var_rnoise, new_model.err, \
+            int_data, int_dq, int_var_poisson, int_var_rnoise, int_err, int_int_times, \
+            opt_slope, opt_sigslope, opt_var_poisson, opt_var_rnoise, opt_yint, opt_sigyint, \
+            opt_pedestal, opt_weights, opt_crmag, actual_segments, actual_CRs
 
-        log.info('Number of leading groups that are flagged as DO_NOT_USE: %s', num_bad_slices)
 
-        # If all groups were flagged, the final group would have been picked up
-        #   in the while loop above, ngroups would have been set to 0, and Nones
-        #   would have been returned.  If execution has gotten here, there must
-        #   be at least 1 remaining group that is not all flagged.
-        last_gdq = groupdq[:, -1, :, :]
-        if np.all(np.bitwise_and(last_gdq, DO_NOT_USE)):
-            nreads -= 1
-            ngroups -= 1
+def ramp_fit_miri_multigroups(data, err, groupdq, cubeshape, imshape, nreads, ngroups):
+    """
+    For MIRI datasets having >1 group, if all pixels in the final group are
+    flagged as DO_NOT_USE, resize the input model arrays to exclude the
+    final group.  Similarly, if leading groups 1 though N have all pixels
+    flagged as DO_NOT_USE, those groups will be ignored by ramp fitting, and
+    the input model arrays will be resized appropriately. If all pixels in
+    all groups are flagged, return None for the models.
 
-            # Check if there are remaining groups before accessing data
-            if ngroups < 1:  # no usable data
-                log.error('2. All groups have all pixels flagged as DO_NOT_USE,')
-                log.error('  so will not process this dataset.')
-                return None, None, None
+    Parameters
+    ----------
+    data: ndarray
+        4-D image cube with dimensions (integrations, groups, rows, columns)
 
-            data = data[:, :-1, :, :]
-            err = err[:, :-1, :, :]
-            groupdq = groupdq[:, :-1, :, :]
-            cubeshape = (nreads,) + imshape
+    err: ndarray
+        Error array
 
-            log.info('MIRI dataset has all pixels in the final group flagged as DO_NOT_USE.')
+    groupdq: ndarray
+        GroupDQ array
 
-        # Next block is to satisfy github issue 1681:
-        # "MIRI FirstFrame and LastFrame minimum number of groups"
-        if ngroups < 2:
-            log.warning('MIRI datasets require at least 2 groups/integration')
-            log.warning('(NGROUPS), so will not process this dataset.')
-            return None, None, None
+    cubeshape: tuple
```

Comment:
Is `cubeshape` deterministically derived from `data.shape` in some way?  From the comment, it looks like `cubeshape = data[0].shape`.  If that's the case, then this doesn't need to be passed separately?